File: Sources/MonitorPresentation/MonitorZoomGeometry.swift
import Foundation

/// Trailing landscape disc vs portrait disc that sits above bottom chrome.
public enum MonitorZoomAttachment: Equatable, Sendable {
    case trailing
    case bottom
}

/// The scale stays clear of a trailing cutout or bottom chrome; its material
/// continues to the attached physical edge as one shape. Coordinates are local
/// to that material's bounds.
public struct MonitorZoomGeometry: Equatable, Sendable {
    public let radius: Double
    public let edgeExtension: Double
    public let attachment: MonitorZoomAttachment

    public init(
        width: Double, height: Double, trailingInset: Double = 0, bottomInset: Double = 0,
        attachment: MonitorZoomAttachment = .trailing
    ) {
        let width = width.isFinite ? max(0, width) : 0
        let height = height.isFinite ? max(0, height) : 0
        self.attachment = attachment
        switch attachment {
        case .trailing:
            edgeExtension = min(width, trailingInset.isFinite ? max(0, trailingInset) : 0)
            let reference = min(
                min(width, height) >= 600 ? 330.0 : 260.0,
                max(120, (height - 24) / 2))
            radius = min(reference * 1.10, height / 2, max(0, width - edgeExtension))
        case .bottom:
            edgeExtension = min(height, bottomInset.isFinite ? max(0, bottomInset) : 0)
            let reference = min(
                min(width, height) >= 600 ? 330.0 : 260.0,
                max(120, (width - 24) / 2))
            radius = min(reference * 1.10, width / 2, max(0, height - edgeExtension))
        }
    }

    public var width: Double {
        attachment == .trailing ? radius + edgeExtension : radius * 2
    }
    public var height: Double {
        attachment == .trailing ? radius * 2 : radius + edgeExtension
    }

    /// Includes the edge extension but excludes the transparent arc corners.
    public func contains(x: Double, y: Double) -> Bool {
        guard radius > 0, x.isFinite, y.isFinite,
            x >= 0, x <= width, y >= 0, y <= height
        else { return false }
        let inExtension = attachment == .trailing ? x >= radius : y >= radius
        if inExtension { return true }
        return pow(x - radius, 2) + pow(y - radius, 2) <= radius * radius
    }

    /// Only the original half-disc can arm a zoom gesture. The flat edge is
    /// excluded because its radial origin is singular at the midpoint.
    public func canStartZoom(x: Double, y: Double) -> Bool {
        guard contains(x: x, y: y) else { return false }
        return attachment == .trailing ? x < radius : y < radius
    }

    /// Original radial mapping; the gesture policy excludes the extension.
    public func angle(x: Double, y: Double) -> Double {
        switch attachment {
        case .trailing:
            return atan2(radius - x, y - radius)
        case .bottom:
            return atan2(radius - y, radius - x)
        }
    }
}

/// One pointer's radial projection, separate from the native edit transaction.
/// The material extension consumes input but cannot arm a gesture. Entering it
/// pauses an armed gesture; the first point back rebases without applying the
/// missing arc. Ordinary disc movement keeps its original angular mapping.
public struct MonitorZoomRadialGesture {
    private let geometry: MonitorZoomGeometry
    public let isArmed: Bool
    private var lastAngle: Double?
    private var accumulatedAngle = 0.0

    public init(geometry: MonitorZoomGeometry, startX: Double, startY: Double) {
        self.geometry = geometry
        isArmed = geometry.canStartZoom(x: startX, y: startY)
        lastAngle = isArmed ? geometry.angle(x: startX, y: startY) : nil
    }

    /// Native drag detectors may discard touch-slop movement before arming.
    /// This advances only the angle reference, never the accumulated output.
    public mutating func rebase(x: Double, y: Double) {
        lastAngle = radialAngle(x: x, y: y)
    }

    public mutating func angleDelta(x: Double, y: Double) -> Double? {
        guard isArmed else { return nil }
        guard let next = radialAngle(x: x, y: y) else {
            lastAngle = nil
            return nil
        }
        guard let previous = lastAngle else {
            lastAngle = next
            return nil
        }
        accumulatedAngle += next - previous
        lastAngle = next
        return accumulatedAngle
    }

    private func radialAngle(x: Double, y: Double) -> Double? {
        guard isArmed, x.isFinite, y.isFinite else { return nil }
        let onScale = geometry.attachment == .trailing ? x < geometry.radius : y < geometry.radius
        guard onScale else { return nil }
        return geometry.angle(x: x, y: y)
    }
}

File: Apps/Android/monitor-ui/src/main/kotlin/com/opencapture/monitorui/MonitorZoomGeometry.kt
package com.opencapture.monitorui

import kotlin.math.atan2

enum class MonitorZoomAttachment { Trailing, Bottom }

/** Mirrors portable MonitorZoomGeometry: an inset scale in one edge-filled material. */
data class MonitorZoomGeometry(
    val radius: Float,
    val edgeExtension: Float,
    val attachment: MonitorZoomAttachment = MonitorZoomAttachment.Trailing,
) {
    val width: Float get() = if (attachment == MonitorZoomAttachment.Trailing) radius + edgeExtension else radius * 2f
    val height: Float get() = if (attachment == MonitorZoomAttachment.Trailing) radius * 2f else radius + edgeExtension

    fun contains(x: Float, y: Float): Boolean {
        if (radius <= 0f || !x.isFinite() || !y.isFinite() ||
            x < 0f || x > width || y < 0f || y > height) return false
        val inExtension = if (attachment == MonitorZoomAttachment.Trailing) x >= radius else y >= radius
        if (inExtension) return true
        return (x - radius) * (x - radius) + (y - radius) * (y - radius) <= radius * radius
    }

    /** Only the original half-disc arms zoom; its singular flat edge is excluded. */
    fun canStartZoom(x: Float, y: Float): Boolean {
        if (!contains(x, y)) return false
        return if (attachment == MonitorZoomAttachment.Trailing) x < radius else y < radius
    }

    fun angle(x: Float, y: Float): Double = if (attachment == MonitorZoomAttachment.Trailing)
        atan2((radius - x).toDouble(), (y - radius).toDouble())
    else atan2((radius - y).toDouble(), (radius - x).toDouble())

    companion object {
        fun layout(
            width: Float, height: Float, trailingInset: Float = 0f, bottomInset: Float = 0f,
            attachment: MonitorZoomAttachment = MonitorZoomAttachment.Trailing,
        ): MonitorZoomGeometry {
            val w = width.takeIf { it.isFinite() }?.coerceAtLeast(0f) ?: 0f
            val h = height.takeIf { it.isFinite() }?.coerceAtLeast(0f) ?: 0f
            return if (attachment == MonitorZoomAttachment.Trailing) {
                val extension = (trailingInset.takeIf { it.isFinite() } ?: 0f).coerceIn(0f, w)
                val reference = minOf(if (minOf(w, h) >= 600f) 330f else 260f,
                    maxOf(120f, (h - 24f) / 2f))
                MonitorZoomGeometry(minOf(reference * 1.10f, h / 2f, w - extension), extension)
            } else {
                val extension = (bottomInset.takeIf { it.isFinite() } ?: 0f).coerceIn(0f, h)
                val reference = minOf(if (minOf(w, h) >= 600f) 330f else 260f,
                    maxOf(120f, (w - 24f) / 2f))
                MonitorZoomGeometry(
                    minOf(reference * 1.10f, w / 2f, h - extension), extension,
                    MonitorZoomAttachment.Bottom,
                )
            }
        }
    }
}

/** Mirrors the Swift pointer policy: extension input pauses, reentry rebases without a jump. */
class MonitorZoomRadialGesture(private val geometry: MonitorZoomGeometry, startX: Float, startY: Float) {
    val isArmed = geometry.canStartZoom(startX, startY)
    private var lastAngle: Double? = if (isArmed) geometry.angle(startX, startY) else null
    private var accumulatedAngle = 0.0

    /** Skip native touch-slop movement without changing accumulated output. */
    fun rebase(x: Float, y: Float) { lastAngle = radialAngle(x, y) }

    fun angleDelta(x: Float, y: Float): Double? {
        if (!isArmed) return null
        val next = radialAngle(x, y)
        if (next == null) { lastAngle = null; return null }
        val previous = lastAngle
        lastAngle = next
        if (previous == null) return null
        accumulatedAngle += next - previous
        return accumulatedAngle
    }

    private fun radialAngle(x: Float, y: Float): Double? {
        if (!isArmed || !x.isFinite() || !y.isFinite()) return null
        val onScale = if (geometry.attachment == MonitorZoomAttachment.Trailing) x < geometry.radius
        else y < geometry.radius
        return if (onScale) geometry.angle(x, y) else null
    }
}
